core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -303,6 +303,11 @@ static HttpRequest of(RequestHeaders headers, Publisher<? extends HttpObject> pu
     static HttpRequest of(HttpRequest request, RequestHeaders newHeaders) {
         requireNonNull(request, "request");
         requireNonNull(newHeaders, "newHeaders");
+        if (request.headers() == newHeaders) {
```

Comment:
You could check this again after unwrapping below.

Reply:
Moved, thank!